### PR TITLE
TimeRangePicker: do not swap time ranges when `to` is before `from`

### LINKED ISSUE #49495: Semi-relative time range breaks since Grafana 8 when 'to' is before 'from'
**Context**:

In response to the blog post [Pro tip: How to use semi-relative time ranges in Grafana](https://grafana.com/blog/2022/02/03/pro-tip-how-to-use-semi-relative-time-ranges-in-grafana/), a user reported an issue, which I was able to reproduce.

There is a dashboard which displays metrics for "the day so far", starting at 9:30am local time. This is achieved with the following time range controls:

* **from**: `now/d+9h+30m`
* **to**: `now+30m`

**What you expected to happen**:

Before Grafana 8, when the dashboard was opened before 9:30am, the graphs would be empty. As soon as the clock hit 9:30am and the dashboard is auto-refreshed, the graphs will display the data between 9:30am and 'now'.

Here is an example:

Screenshot 2022-05-24 09-40-12 now before start

Screenshot 2022-05-24 09-41-26 now after start

**What happened**:

Since Grafana 8, when the dashboard is opened before 9:30am, the time controls are reversed. On 2022-05-24 at 9:25am, the displayed time range is `2022-05-24 09:25:00` to `2022-05-25 09:30:00`.

Screenshot 2022-05-24 09-25-52 before 9:30

After 9:30am passes and as the dashboard auto-refreshes, the dashboard time controls remain reversed.

Screenshot 2022-05-24 09-31-26 after 9:30

Refreshing the page on the browser controls fixes the issue:

Screenshot 2022-05-24 09-32-23 after refresh

**Environment**:
- Grafana version: v8.5.2
- Data source type & version: Prometheus
- OS Grafana is installed on: Ubuntu 20.04
- User OS & Browser: Firefox v100
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by eledobleefe**
**What is this feature?**

It undoes the following code change and does not swap time ranges:

```jsx 
if (range.to.isBefore(range.from)) {
      this.setTime(
        {
          from: range.raw.to,
          to: range.raw.from,
        },
        false
      );
    }
```

**Why do we need this feature?**

To avoid swapping time ranges when the `to` value is before the `from` value and also to notify the user that this is kind of an invalid value that can lead to not showing/returning any data.

**Who is this feature for?**

Everybody

**Which issue(s) does this PR fix?**:

Fixes #49495 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by eledobleefe** [comment]
We have proceeded in the following way:
1.- Get rid of the piece of code that swapped the time ranges
2.- Trigger an AppEvent to show a Toast. This will notify the user that this time range can lead to not returning/showing data. This change is mainly focused on the time range being changed in the URL params.
Captura de pantalla 2023-11-30 a las 16 39 44

3.- With the current code (previous to these changes) the user couldn't set a `to` value that is before the `from` one. We keep this behaviour.
Captura de pantalla 2023-11-30 a las 16 39 57

File: public/app/features/dashboard/services/TimeSrv.ts [hunk 1]
```diff
@@ -77,13 +78,14 @@ export class TimeSrv {
...
-          to: range.raw.from,
-        },
-        false
-      );
+      const appEvents = getAppEvents();
+      appEvents.publish({
+        type: AppEvents.alertWarning.name,
+        payload: [
+          'Invalid time range',
+          'This dashboard may not show any data as the time range is invalid: `to` value is before `from`.',
```
Comment (JoaoSilvaGrafana) [inline public/app/features/dashboard/services/TimeSrv.ts:86]: My only comment is that I think the warning should say something similar to the error

```suggestion
          'This dashboard may not show any data as the time range is invalid: "From" value is after "To".',
```

**Comment by joshhunt** [comment]
Good catch about needing to show warnings outside the time picker (like when direct linked)! I had not considered that before.

Hmmm. I'm not enthusastic about using toasts for this. They quickly dismiss so it can be easy to not see them.

I would propose a more persistent warning - perhaps changing the time picker icon to a warning? (excuse my _very_ crude drawings)

6565_2023-11-30-17-46_firefox

I'm also not sure using the field error is appropriate - this isn't a strict error the user _must_ act on, instead it's a warning. Sometimes (such as the original issue) the selected time range is actually correct, and the warning is just informational. Perhaps we could should a warning seperated from the fields?

6566_2023-11-30-17-47_firefox

File: public/app/features/dashboard/services/TimeSrv.ts [hunk 2]
```diff
@@ -77,13 +78,14 @@ export class TimeSrv {
...
 
     if (range.to.isBefore(range.from)) {
-      this.setTime(
-        {
-          from: range.raw.to,
-          to: range.raw.from,
-        },
-        false
-      );
+      const appEvents = getAppEvents();
```
Comment (joshhunt) [inline public/app/features/dashboard/services/TimeSrv.ts:81]: Worth checking as well whether this works with Dashboards-as-scenes?

Enable the `scenes=true` feature flag and click the "View as Scene" button.

**Comment by eledobleefe** [comment]
> Good catch about needing to show warnings outside the time picker (like when direct linked)! I had not considered that before.
> 
> Hmmm. I'm not enthusastic about using toasts for this. They quickly dismiss so it can be easy to not see them.
> 
> I would propose a more persistent warning - perhaps changing the time picker icon to a warning? (excuse my _very_ crude drawings)
> 
> 6565_2023-11-30-17-46_firefox
> 
> I'm also not sure using the field error is appropriate - this isn't a strict error the user _must_ act on, instead it's a warning. Sometimes (such as the original issue) the selected time range is actually correct, and the warning is just informational. Perhaps we could should a warning seperated from the fields?
> 
> 6566_2023-11-30-17-47_firefox

Hi @joshhunt,
Thanks for your input and also for your drawings 😉 
I agree the Toast might not be a good notification as it disappears in a short period of time and really appreciate your proposal. I'll reach out to you when it's done.
Regarding this behaviour:
Captura de pantalla 2023-11-30 a las 16 39 57
This is the current one. I think we are using an error appearance because the user cannot apply this time range so an action from the user is needed to continue (changing the time range). In this case, I'm not sure we should make a change 🤔

**Comment by joshhunt** [comment]
Hmmm - I wonder what happens if you try to set the time range specified in https://github.com/grafana/grafana/issues/49495 before 9:30am? Does it prevent you from saving the dashboard because from is before to?

I agree with you - lets not change that existing behaviour right now - but it's something to think about.

**Comment by eledobleefe** [comment]
> Hmmm - I wonder what happens if you try to set the time range specified in #49495 before 9:30am? Does it prevent you from saving the dashboard because from is before to?
> 
> I agree with you - lets not change that existing behaviour right now - but it's something to think about.

I checked it and I could not set a time range if the `From` value was after the `To` value so you need to change it, otherwise it won't be applied. The tricky part is that the user can set a `From` after the `To` using the URL params, so that is what we are managing when showing the Toast or, in a while, a warning icon in the TimeRangePicker.
I'll check it again anyway after applying the changes I'm working on 😉

**Comment by eledobleefe** [comment]
@joshhunt, I worked with the `ToolbarButton` instead of the Toast. We can show that something is wrong with the `TimeRangePicker` using one of these appearances, but not sure if they're right or not, so will ask @staton-hyse11 for his input too:
1.- Option 1: The icon changes to the 'exclamation-triangle' one. This might not be enough highlighted to be noticed 🤔 
Captura de pantalla 2023-12-01 a las 12 14 36
2.- Option 2: Set the `variant` to `active` on the `ToolbarButton`.
Captura de pantalla 2023-12-01 a las 12 06 55

3.- Option 3: Set the `variant` to `destructive` on the `ToolbarButton`. After looking at some documentation, I would go for this one because this is meant 'to attract the user's attention'. After all, if the user does not change the time range the dashboard will not work as expected.
Captura de pantalla 2023-12-01 a las 12 07 44

I'd appreciate any feedback to come up with the best option.

**Comment by joshhunt** [comment]
@eledobleefe Nice stuff!

I think this is tricky because it's not a hard error that _requires_ user attention - it's just a warning that is valid to ignore in some scenarios (like before 9:30am).

In my mind I imagined just an orange ⚠️  icon in the toolbar, but i appreciate I don't think we can specify that at the moment.

File: public/app/features/dashboard/services/TimeSrv.ts (hunk 2 above)
Comment (eledobleefe) [reply public/app/features/dashboard/services/TimeSrv.ts]: As we changed this to not use a `Toast`, this compatibility check is no longer needed.

**Review by JoaoSilvaGrafana — APPROVED** [review]
I like this solution, this is probably an edge case so this is simple enough but should be clear that something is wrong. The key part is that we deleted that piece of code that reverts the time range 👍

**Comment by eledobleefe** [comment]
> I like this solution, this is probably an edge case so this is simple enough but should be clear that something is wrong. The key part is that we deleted that piece of code that reverts the time range 👍

Agreed. Btw, did you test it through the instance?

**Comment by JoaoSilvaGrafana** [comment]
> Agreed. Btw, did you test it through the instance?

I didn't when I approved, I just pulled the branch. But I did test it now with that too! 🎉
